File: go/111.go

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 5 star, 按层次遍历，当前层的节点如果有一个是叶子，则当前层的深度即是最小深度

func minDepth(root *TreeNode) int {
	if root == nil {return 0}
	if root.Right == nil && root.Left == nil {return 1}
	queue := []*TreeNode{root}
	depth := 0
	for len(queue) > 0 {
		depth += 1
		nextLevel := []*TreeNode{}
		for _, node := range queue {
			if node.Left == nil && node.Right == nil {return depth}
			if node.Left != nil {nextLevel = append(nextLevel, node.Left)}
			if node.Right != nil {nextLevel = append(nextLevel, node.Right)}
		}
		queue = nextLevel
	}
	return depth


}

File: v2/111.py





# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        # 6 star, 分类讨论的情形比较巧妙，一开始做错了
        # todo, 需要掌握
        if not root:
            return 0
        return self._minDepth(root)

    def _minDepth(self, node):
        if not node.left and not node.right:
            return 1
        elif node.left and node.right:
            return min(self._minDepth(node.left), self._minDepth(node.right)) + 1
        elif node.left:
            return self._minDepth(node.left) + 1
        else:
            return self._minDepth(node.right) + 1